clusterctl - Allow for disabling pivoting of cluster

detiber opened the pull request:
**What this PR does / why we need it**:
- Add --no-pivot-cluster flag to disable pivoting of the cluster

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #450

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
- clusterctl adds the --no-pivot-cluster flag to 'clusterctl create cluster' that allows for disabling pivoting of cluster-api components.
```

ingvagabund commented on clusterctl/clusterdeployer/clusterdeployer.go:
```diff
@@ -148,39 +151,50 @@ func (d *ClusterDeployer) Create(cluster *clusterv1.Cluster, machines []*cluster
 		return fmt.Errorf("unable to create node machines: %v", err)
 	}
 
 	if d.addonComponents != "" {
```
What is purpose of the addon components? Do they extend the cluster API stack? Checking:
* https://github.com/kubernetes-sigs/cluster-api-provider-gcp/blob/master/clusterctl/examples/google/addons.yaml.template
* https://github.com/roberthbailey/cluster-api-provider-vsphere/blob/master/clusterctl/examples/vsphere/addons.yaml.template

I can't decide if they are meant for the external of the internal cluster.

roberthbailey replied:
They are meant for the internal cluster. The provider components are the cluster api stack that is applied to the internal and external cluster. The addon components are things we "add on" to the internal cluster in addition to the core k8s control plane as part of the baseline installation.

ingvagabund commented on clusterctl/clusterdeployer/clusterdeployer.go:
```diff
@@ -148,39 +151,50 @@ func (d *ClusterDeployer) Create(cluster *clusterv1.Cluster, machines []*cluster
+		if err := d.updateClusterEndpoint(internalClient, provider); err != nil {
+			return fmt.Errorf("unable to update internal cluster endpoint: %v", err)
+		}
+	} else {
```
If the `pivotCluster` is false, the `cleanupExternalCluster` needs to be set to `false`. Otherwise, we will provision the cluster API stack and then remove it.

detiber replied:
This has been addressed here: https://github.com/kubernetes-sigs/cluster-api/blob/b546f32729155b36ae1068b3fef8671d98e209a5/clusterctl/clusterdeployer/clusterdeployer.go#L247

roberthbailey commented:
/assign @ingvagabund

ingvagabund commented:
@roberthbailey unfortunately I am not in the `kubernetes-sigs` org. What's the typical trial to get to the org? Just request for it?

roberthbailey commented:
Yes. Since you are a member of the kubernetes org, you have implicit membership in the kubernetes-sigs org as well but you need to request it. 

From https://github.com/kubernetes/community/blob/master/community-membership.md#kubernetes-ecosystem

> We are currently working on automation that would transfer membership in the Kubernetes organization to any related orgs automatically, but such is not the case currently.

detiber commented:
@ingvagabund I believe I've addressed all of the comments not already answered. I agree about the unfortunate naming of external/internal, but believe that would be better handled through a separate issue/PR.

ingvagabund commented:
@roberthbailey I am in the sig now, thanks for the hints
/assign ingvagabund

ingvagabund commented:
@detiber thanks for addressing the comments. Looks good now. Lemme just build and run the clusterctl command locally.

ingvagabund commented:
@detiber lgtm. Can you squash the commits?

detiber commented:
@ingvagabund squashed. I left the version bump for the api server image as a separate commit since it's not directly related to the change, thanks for the review!

ingvagabund commented:
/lgtm

ingvagabund commented:
/approve

ingvagabund commented:
@medinatiger can you please approve?

roberthbailey commented:
I was looking at this today and I had a couple of questions that I wanted to discuss in the wg meeting.

roberthbailey commented:
/assign @roberthbailey

roberthbailey commented:
@detiber - please ping this PR once you've updated it to reflect the discussion during the WG meeting today.

detiber commented:
Rebased, but still need to address the default behavior as discussed in the cluster-api meeting.

roberthbailey commented:
/lgtm
/approve
/hold

Holding for a bit in case other folks want to comment before the merge bot takes over.

frodenas commented:
@roberthbailey There haven't been any comments in the last 28 days, can this be merged now?

roberthbailey commented:
Yes, although it now needs to be rebased.

detiber commented:
@frodenas @roberthbailey I can rebase this next week (and after #494) is merged

roberthbailey commented:
@detiber - let me know when this has been rebased and is ready for review.

detiber commented:
Closing in favor of the phases support.
